File: python/heapsort.py
def heapsort(numbers):
  count = len(numbers)
  heapify(numbers, count)

  end = count - 1
  while(end > 0):
    numbers[end], numbers[0] = numbers[0], numbers[end]
    end -= 1
    sift_down(numbers, 0, end)

  return numbers

def heapify(numbers, count):
  start = (count - 1)//2 # starting parent is here.
  while(start >= 0 ):
    sift_down(numbers, start, count-1)
    start -= 1

def sift_down(numbers, start, end):
  root = start
  left_child = (2 * root) + 1
  while(left_child <= end):
    child = left_child
    swap = root

    if(numbers[child] > numbers[swap]):
      swap = child

    if(child + 1 <= end and numbers[child+1] > numbers[swap]):
      swap = child + 1

    if(swap == root):
      return

    else:
      swapped_number = numbers[swap]
      numbers[swap] = numbers[root]
      numbers[root] = swapped_number
      root = swap
      left_child = (2 * root) + 1


numbers = [3,2,1,5,4,0,9,6,7,8,8,12,3,4,1,2,5,6,3,1,-3,-5,-6]

print(heapsort(numbers))

File: javascript/heapsort.js
function heapsort (numbers) {
  // First we heapify the numbers
  const size = numbers.length
  heapify(numbers, size)
  var end = size - 1

  while(end > 0){
    let tmp = numbers[0]
    numbers[0] = numbers[end]
    numbers[end] = tmp
    end -= 1
    siftDown(numbers, 0, end)
  }
  return numbers
}

function heapify (numbers, size) {
  start = Math.floor((size -1) /2)
  while(start>= 0){
    siftDown(numbers, start, size -1)
    start -= 1
  }
}

function siftDown(numbers, start, end) {
  var root = start
  var swap = root
  var leftChild = (root * 2) + 1
  while(leftChild <= end){
    if(numbers[swap] < numbers[leftChild]){
      swap = leftChild
    }

    if(leftChild + 1 <= end && numbers[swap] < numbers[leftChild + 1]){
      swap = leftChild + 1
    }

    if(swap === root) {
      return
    } else {
      let tmp = numbers[root]
      numbers[root] = numbers[swap]
      numbers[swap] = tmp
      root = swap
      leftChild = (root * 2) + 1
      console.log(numbers)
    }
  }
}


console.log(heapsort([1,4,2,3,7,6,9,9,8,0]))
